service/keychain/local_keychain/local_keychain.go

```diff
@@ -0,0 +1,130 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package local_keychain
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/awslabs/soci-snapshotter/proto"
+	"github.com/awslabs/soci-snapshotter/service/resolver"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/reference"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+)
+
+var (
+	localKeychainPort = flag.Int("local_keychain_port", 0,
+		"Port on which to expose the local_keychain gRPC service that accepts username/password credentials for private images. If 0, the local_keychain service is not started/exposed.")
+)
+
+type credentials struct {
+	username   string
+	password   string
+	expiration *time.Time
+}
+
+type keychain struct {
+	cache map[string]credentials
+	proto.UnimplementedLocalKeychainServer
```

Comment:
I don't see why you need this.

Reply:
Added it to make the grpc_cli work (without this, `grpc_cli ls ...` failed), but I can remove it.

Reply:
Can you try again? That doesn't make much sense. 
All proto.UnimplementedLocalKeychainServer does is provide no-op implementations for all RPC methods in the proto and in your case you already have an implementation for `PutCredentials` so having UnimplementedLocalKeychainServer here is a no-op.
But if it actually fixes something feel free to keep it, it's not a big deal.

Reply:
Looks like I was mistaken, `grpc_cli ls` and `grpc_cli call` work fine without this. Not sure why I added it in the first place, but I'll leave it out.